File: autovalet_navigation/src/move_base_listener.cpp
#include <move_base_listener.h>
#include <geometry_msgs/TransformStamped.h>


void getCurrentPose(geometry_msgs::PoseStamped &pose)
{
    // get the current transform from map to base_link and store that in "pose"
    geometry_msgs::TransformStamped transform;
    tf2_ros::Buffer tf_buffer;
    tf2_ros::TransformListener tf_listener(tf_buffer);
    transform = tf_buffer.lookupTransform("icp_odom","map",ros::Time(0), ros::Duration(1.0));
    
    pose.header = transform.header;
    pose.pose.position.x = transform.transform.translation.x;
    pose.pose.position.y = transform.transform.translation.y;
    pose.pose.position.z = transform.transform.translation.z;
    pose.pose.orientation = transform.transform.rotation;
}

// checks if poses are equal. this might need to be changed later to account for tolerances...
bool arePosesClose(geometry_msgs::PoseStamped pose_1, geometry_msgs::PoseStamped pose_2)
{
    bool result = true;

    if (result && (pose_1.pose.position.x != pose_2.pose.position.x))
    {
        result = false;
    }
    if (result && (pose_1.pose.position.y != pose_2.pose.position.y))
    {
        result = false;
    }
    return result;
}

// constructor. saves topic names
MoveBaseListener::MoveBaseListener(std::string goal_topic, std::string status_topic)
{
    m_goal_topic = goal_topic;
    m_status_topic = status_topic;
}

void MoveBaseListener::init() 
{
    // subscribe to the goal topic and the status topic
    m_goal_sub = m_nh.subscribe(m_goal_topic, 1000, &MoveBaseListener::goalCallback, this);
    m_status_sub = m_nh.subscribe(m_status_topic, 1000, &MoveBaseListener::statusCallback, this);

    // start off by waiting for a goal pose
    m_plan_state = Waiting;
    
    printState();
}

void MoveBaseListener::goalCallback(const geometry_msgs::PoseStamped& goal_msg)
{
    // move to planning state
    m_plan_state = Planning;

    // store the goal pose
    m_goal_pose = goal_msg;

    // get the current pose and store that as the start
    getCurrentPose(m_start_pose);

    printState();

    ROS_WARN("Goal:");
    ROS_WARN("Position");
    ROS_WARN("  x: %f",m_goal_pose.pose.position.x);
    ROS_WARN("  y: %f",m_goal_pose.pose.position.y);
    ROS_WARN("  z: %f",m_goal_pose.pose.position.z);
    ROS_WARN("Orientation");
    ROS_WARN("  x: %f",m_goal_pose.pose.orientation.x);
    ROS_WARN("  y: %f",m_goal_pose.pose.orientation.y);
    ROS_WARN("  z: %f",m_goal_pose.pose.orientation.z);
    ROS_WARN("  w: %f",m_goal_pose.pose.orientation.w);
}

void MoveBaseListener::statusCallback(const actionlib_msgs::GoalStatusArray::ConstPtr& status_msg)
{
    // goal status codes 
    // Documentation: http://docs.ros.org/jade/api/actionlib_msgs/html/msg/GoalStatus.html
    static const int ACTIVE = 1;
    static const int SUCCEEDED = 3;
    static const int ABORTED = 4;

    // check to make sure that at least one message has been sent (prevents seg faults)
    if (!status_msg->status_list.empty())
    {
        // Documentation: http://docs.ros.org/jade/api/actionlib_msgs/html/msg/GoalStatus.html
        actionlib_msgs::GoalStatus current_status = status_msg->status_list.back();

        // if we're in the waiting state and haven't received a new goal, no need to update anything
        if (current_status.status != ACTIVE && m_plan_state == Waiting) 
        {
            return;
        }
        
        // Process status code read from the /move_base/status topic
        if (current_status.status == ACTIVE)
        {
            // if were in the waiting state and we've received a goal, move to planning
            if (m_plan_state == Waiting)
            {
                m_plan_state = Planning;
            } 

            // if we're in the planning state and we've started moving, then we must be in the
            // executing state. if we haven't started moving, return so we don't print the state
            // again
            else if (m_plan_state == Planning)
            {
                geometry_msgs::PoseStamped current_pose;
                getCurrentPose(current_pose);

                if (!arePosesClose(current_pose,m_start_pose))
                {
                    m_plan_state = Executing;
                } else {
                    return;
                }
            } 
            
            // if we're in the executing state, don't do anything (we've already printed that we're in this state)
            else if (m_plan_state == Executing)
            {
                return;
            }
        } 

        else if (current_status.status == SUCCEEDED)
        {
            // if we're already in the waiting state, return so we don't print that state again
            if (m_plan_state == Waiting)
            {
                return;
            } else{
                // print success and then shift to the waiting state
                m_plan_state = Success;
                printState();
                m_plan_state = Waiting;
            } 
        }
        else if (current_status.status == ABORTED && m_plan_state != Waiting) //4
        {
            // if we're in the waiting state, return so we don't print state again
            if (m_plan_state == Waiting)
            {
                return;
            } else {
                // print fail and then shift to the waiting state
                m_plan_state = Fail;
                printState();
                m_plan_state = Waiting;
            }
        } 
    } else {
        return;
    }

    // print the current state. this fxn should only get called if we have just changed states
    printState();
}


void MoveBaseListener::printState()
{
    std::string output = "Current State: ";
    ROS_WARN("---------");
    switch(m_plan_state)
    {
        case Waiting:
            output += "WAITING FOR NEXT GOAL";
            break;
        case Planning:
            output += "PLANNING PATH TO GOAL";
            break;
        case Executing:
            output += "EXECUTING PLAN";
            break;
        case Fail:
            output += "FAILED TO FIND A PLAN";
            break;
        case Success:
            output += "SUCCESSFULLY REACHED GOAL";
            break;
    }
    ROS_WARN_STREAM(output);
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "planner");

    std::string goal_topic = "/move_base_simple/goal";
    std::string status_topic = "/move_base/status";

    MoveBaseListener listener(goal_topic, status_topic);
    listener.init();
    ros::spin();

    return 0;
}

File: autovalet_navigation/scripts/autovalet_navigation/move_base_listener.py
#!/usr/bin/env python

import numpy as np

# Include ROS libs
import rospy
import tf2_ros
from tf2_geometry_msgs import do_transform_pose

# Include messages
from geometry_msgs.msg import PoseStamped, Quaternion, Pose
from tf.transformations import quaternion_matrix, quaternion_from_matrix
from actionlib_msgs.msg import GoalStatusArray

def getCurrentPose():

    tf_buffer      = tf2_ros.Buffer(rospy.Duration(1200.0)) # Length of tf2 buffer (?)
    tf_listener    = tf2_ros.TransformListener(tf_buffer)


    transform = tf_buffer.lookup_transform("icp_odom", # target_frame_id
                                "map", #source frame
                                rospy.Time(0), #get the tf at first available time
                                rospy.Duration(1.0)) #timeout after 1

    current_pose = PoseStamped()
    current_pose.header = transform.header
    current_pose.pose.position.x = transform.transform.translation.x
    current_pose.pose.position.y = transform.transform.translation.y
    current_pose.pose.position.z = transform.transform.translation.z
    current_pose.pose.orientation = transform.transform.rotation

    return current_pose

def arePosesClose(pose1, pose2):
    result = True

    if result and pose1.pose.position.x != pose2.pose.position.x:
        result = False
    if result and pose1.pose.position.y != pose2.pose.position.y:
        result = False

    return result

class MoveBaseState:
    Waiting = 0
    Planning = 1
    Executing = 2
    Fail = 3
    Success = 4

class MoveBaseListener:
    def __init__(self, debug=True):
        self.goal_sub       = rospy.Subscriber("/move_base_simple/goal", PoseStamped, self.goalCallback)
        self.status_sub     = rospy.Subscriber("/move_base/status", GoalStatusArray, self.MBstatusCallback)
        
        self.goal_pose      = PoseStamped()
        self.start_pose     = PoseStamped()

        self.plan_state     = MoveBaseState.Waiting

        self.debug          = debug

    def getState(self):
        return self.plan_state 

    def printState(self):
        rospy.logwarn("---------")
        output = "Current State: "

        if self.plan_state == MoveBaseState.Waiting:
            output += "WAITING FOR NEXT GOAL"
        elif self.plan_state == MoveBaseState.Planning:
            output += "PLANNING PATH TO GOAL"
        elif self.plan_state == MoveBaseState.Executing:
            output += "EXECUTING PLAN"
        elif self.plan_state == MoveBaseState.Fail:
            output += "FAILED TO FIND A PLAN"
        elif self.plan_state == MoveBaseState.Success:
            output += "SUCCESSFULLY REACHED GOAL"

        rospy.logwarn(output)

    def getGoal():
        return self.goal_pose

    def goalCallback(self,goal_msg):
        
        self.plan_state = MoveBaseState.Planning

        self.goal_pose = goal_msg

        self.start_pose = getCurrentPose()

        if self.debug:
            self.printState()

            rospy.logwarn("Goal:")
            rospy.logwarn("Position")
            rospy.logwarn("  x: %f",self.goal_pose.pose.position.x)
            rospy.logwarn("  y: %f",self.goal_pose.pose.position.y)
            rospy.logwarn("  z: %f",self.goal_pose.pose.position.z)
            rospy.logwarn("Orientation");
            rospy.logwarn("  x: %f",self.goal_pose.pose.orientation.x)
            rospy.logwarn("  y: %f",self.goal_pose.pose.orientation.y)
            rospy.logwarn("  z: %f",self.goal_pose.pose.orientation.z)
            rospy.logwarn("  w: %f",self.goal_pose.pose.orientation.w)

    def MBstatusCallback(self,status_array_msg):

        # print(status_array_msg.__dict__)
        ACTIVE = 1
        SUCCEEDED = 3
        ABORTED = 4

        if status_array_msg.status_list:
            current_status = status_array_msg.status_list[-1]

            if current_status.status != ACTIVE and self.plan_state == MoveBaseState.Waiting:
                return

            if current_status.status == ACTIVE:
                if self.plan_state == MoveBaseState.Waiting:
                    self.plan_state = MoveBaseState.Planning
            
                elif self.plan_state == MoveBaseState.Planning:
                    current_pose = getCurrentPose()

                    if not arePosesClose(current_pose,self.start_pose):
                        self.plan_state = MoveBaseState.Executing
                    else:
                        return
                elif self.plan_state == MoveBaseState.Executing:
                    return

            elif current_status.status == SUCCEEDED:
                if self.plan_state == MoveBaseState.Waiting:
                    return
                else:
                    self.plan_state = MoveBaseState.Success
                    if self.debug:
                        self.printState()
                    self.plan_state = MoveBaseState.Waiting
            
            elif current_status.status == ABORTED and self.plan_state != MoveBaseState.Waiting:
                if self.plan_state == MoveBaseState.Waiting:
                    return

                else:
                    self.plan_state = MoveBaseState.Fail
                    if self.debug:
                        self.printState()
                    self.plan_state = MoveBaseState.Waiting

        else:
            return

        if self.debug:
            self.printState()


if __name__ == '__main__':

    rospy.init_node('planner')

    mbl = MoveBaseListener()

    try:
        rospy.spin()
    except KeyboardInterrupt:
        print("Shutting down")